### ISSUE TITLE
Negated PatternMatcher causes StackOverflowError always

### ISSUE THREAD

**Issue by shevek** [gradle/gradle#11784]
Builds sometimes fail with:
```
        at org.gradle.api.internal.file.pattern.PatternMatcher$3.test(PatternMatcher.java:52)
        at org.gradle.api.internal.file.pattern.PatternMatcher$3.test(PatternMatcher.java:52)
        at org.gradle.api.internal.file.pattern.PatternMatcher$3.test(PatternMatcher.java:52)
        at org.gradle.api.internal.file.pattern.PatternMatcher$3.test(PatternMatcher.java:52)
        at org.gradle.api.internal.file.pattern.PatternMatcher$3.test(PatternMatcher.java:52)
        at org.gradle.api.internal.file.pattern.PatternMatcher$3.test(PatternMatcher.java:52)
        at org.gradle.api.internal.file.pattern.PatternMatcher$3.test(PatternMatcher.java:52)
        at org.gradle.api.internal.file.pattern.PatternMatcher$3.test(PatternMatcher.java:52)
        at org.gradle.api.internal.file.pattern.PatternMatcher$3.test(PatternMatcher.java:52)
        at org.gradle.api.internal.file.pattern.PatternMatcher$3.test(PatternMatcher.java:52)

```

### Current Behavior

https://github.com/gradle/gradle/blob/master/subprojects/files/src/main/java/org/gradle/api/internal/file/pattern/PatternMatcher.java#L55
This `PatternMatcher.this` refers to the self (inner) instance, not the outer instance.

Looks like a simple mistake which very rarely triggers, except when using incremental build with build cache enabled. However, when it does trigger, the project becomes unbuildable, as compileJava always throws StackOverflowException.

**Comment by shevek** [comment]
More useful stack context:
```
Caused by: com.google.common.util.concurrent.ExecutionError: java.lang.StackOverflowError
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2047)
        at com.google.common.cache.LocalCache.get(LocalCache.java:3851)
        at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4713)
        at org.gradle.api.tasks.util.internal.CachingPatternSpecFactory$CachingSpec.isSatisfiedBy(CachingPatternSpecFactory.java:72)
        at org.gradle.api.tasks.util.internal.CachingPatternSpecFactory$CachingSpec.isSatisfiedBy(CachingPatternSpecFactory.java:61)
        at org.gradle.api.specs.AndSpec.isSatisfiedBy(AndSpec.java:48)
        at org.gradle.api.internal.file.collections.DefaultDirectoryWalker.shouldVisit(DefaultDirectoryWalker.java:51)
        at org.gradle.api.internal.file.collections.DefaultDirectoryWalker$PathVisitor.preVisitDirectory(DefaultDirectoryWalker.java:88)
        at org.gradle.api.internal.file.collections.DefaultDirectoryWalker$PathVisitor.preVisitDirectory(DefaultDirectoryWalker.java:66)
        at org.gradle.api.internal.file.collections.DefaultDirectoryWalker.walkDir(DefaultDirectoryWalker.java:60)
        at org.gradle.api.internal.file.collections.DirectoryFileTree.walkDir(DirectoryFileTree.java:160)
        at org.gradle.api.internal.file.collections.DirectoryFileTree.visitFrom(DirectoryFileTree.java:138)
        at org.gradle.api.internal.file.collections.DirectoryFileTree.visit(DirectoryFileTree.java:123)
        at org.gradle.api.internal.file.collections.FileTreeAdapter.visit(FileTreeAdapter.java:119)
        at org.gradle.api.internal.file.AbstractFileTree.getFiles(AbstractFileTree.java:58)
        at org.gradle.api.internal.file.CompositeFileCollection.getFiles(CompositeFileCollection.java:76)
        at org.gradle.api.internal.file.CompositeFileCollection.getFiles(CompositeFileCollection.java:54)
        at org.gradle.api.internal.tasks.compile.incremental.recomp.AbstractRecompilationSpecProvider.deleteStaleFilesIn(AbstractRecompilationSpecProvider.java:51)
        at org.gradle.api.internal.tasks.compile.incremental.recomp.JavaRecompilationSpecProvider.initializeCompilation(JavaRecompilationSpecProvider.java:76)
        at org.gradle.api.internal.tasks.compile.incremental.SelectiveCompiler.execute(SelectiveCompiler.java:73)
        at org.gradle.api.internal.tasks.compile.incremental.SelectiveCompiler.execute(SelectiveCompiler.java:36)
        at org.gradle.api.internal.tasks.compile.incremental.IncrementalResultStoringCompiler.execute(IncrementalResultStoringCompiler.java:60)
        at org.gradle.api.internal.tasks.compile.incremental.IncrementalResultStoringCompiler.execute(IncrementalResultStoringCompiler.java:44)
        at org.gradle.api.internal.tasks.compile.CompileJavaBuildOperationReportingCompiler$2.call(CompileJavaBuildOperationReportingCompiler.java:59)
        at org.gradle.api.internal.tasks.compile.CompileJavaBuildOperationReportingCompiler$2.call(CompileJavaBuildOperationReportingCompiler.java:51)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
        at org.gradle.api.internal.tasks.compile.CompileJavaBuildOperationReportingCompiler.execute(CompileJavaBuildOperationReportingCompiler.java:51)
        at org.gradle.api.tasks.compile.JavaCompile.performCompilation(JavaCompile.java:158)
        at org.gradle.api.tasks.compile.JavaCompile.compile(JavaCompile.java:126)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:103)
        at org.gradle.api.internal.project.taskfactory.IncrementalTaskInputsTaskAction.doExecute(IncrementalTaskInputsTaskAction.java:47)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:42)
        at org.gradle.api.internal.project.taskfactory.AbstractIncrementalTaskAction.execute(AbstractIncrementalTaskAction.java:25)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:28)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$5.run(ExecuteActionsTaskExecuter.java:476)
... caused by SOE ad nauseam
```

**Comment by shevek** [comment]
Good workaround: `tasks.withType(JavaCompile) { options.incremental = false; }`
Non-workaround: `--no-build-cache`

Weirdly, it seems that making some nontrivial changes to the Java code makes this error appear or vanish. I have no idea why. Moving a default method into an abstract class made the code compile OK. Moving it back to the interface makes the compile throw. Making compile be non-incremental makes it pass again.

**Comment by big-guy** [comment]
Which version of Gradle are you using?  This sounds like the same problem as https://github.com/gradle/gradle/issues/10329 which was fixed in 5.6.3 and 6.0.  The line numbers line up for that and are unrelated to negate().  You can see the previous code here: https://github.com/gradle/gradle/blob/2d5aab47a3e4c181af3607860b1e1c5be7ea8166/subprojects/files/src/main/java/org/gradle/api/internal/file/pattern/PatternMatcher.java#L52

> This PatternMatcher.this refers to the self (inner) instance, not the outer instance.

I don't think this is true.

Here's the full block on master:
https://github.com/gradle/gradle/blob/5f65be4aa20ff079c25c0af9e1ad5ebaafbe0eac/subprojects/files/src/main/java/org/gradle/api/internal/file/pattern/PatternMatcher.java#L52-L57

`PatternMatcher.this` is the outer "this" for the pattern being negated. You can confirm this by adding a field to the anonymous class and see that it's reachable from `this.` but not `PatternMatcher.this.`  This could be rewritten to use a named static inner class like the And/Or cases to be consistent.
